### PR TITLE
Adds `CombinedOptimizer`

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

Adds a new `CombinedOptimizer` utility, which is useful for the increasingly-popular "architecture-aware optimizers", such as [Muon](github.com/KellerJordan/Muon).

This PR targets the v2.0 refactor branch, so this should only be merged after #1235 .

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/optim/combined_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        If ``optimizers`` is empty.
+
+    Notes
+    -----
+    * **Parameter Groups**: The ``param_groups`` attribute aggregates parameter
+      groups from all underlying optimizers, making this wrapper compatible with
+      learning rate schedulers.
+    * **Closure Behavior**: When ``step()`` is called with a closure, the closure
+      is evaluated once (not once per optimizer) and the resulting loss is passed
+      to all optimizers. This differs from calling each optimizer's step separately.
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/optim/combined_optimizer.py]: Docs fixed in [c657c5b](https://github.com/NVIDIA/physicsnemo/pull/1241/commits/c657c5b2f95e1557755cdf8b6d27a059f24dc33f)

File: test/optim/test_combined_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+            combined_optimizer.zero_grad()
+            loss = model(torch.randn(1, 10)).sum()
+            loss.backward()
+            return loss
+
+        combined_optimizer.step(closure)
+
+        # CombinedOptimizer has 2 optimizers (SGD, Adam), both support closure
+        # So closure should be called at least twice (once per optimizer)
+        assert call_count >= 2, f"Closure called {call_count} times, expected >= 2"
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply test/optim/test_combined_optimizer.py:196]: Docs fixed in [c657c5b](https://github.com/NVIDIA/physicsnemo/pull/1241/commits/c657c5b2f95e1557755cdf8b6d27a059f24dc33f)

File: physicsnemo/optim/combined_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+        ### Setup step functions (optionally compiled)
+        if torch_compile_kwargs is None:
+            self.step_fns: list[Callable] = [opt.step for opt in optimizers]
+        else:
+            self.step_fns: list[Callable] = [
+                torch.compile(opt.step, **torch_compile_kwargs) for opt in optimizers
+            ]
+
+    def zero_grad(self, *args, **kwargs) -> None:
```
Comment (coreyjadams) [inline physicsnemo/optim/combined_optimizer.py]: Upstream interface accepts only `set_to_none=True` here, and not other parameters.  I know if we passed it, that would succeed, but unless there are optimizers accepting other vales for `zero_grad` I think we should stick to `set_to_none`.

File: physicsnemo/optim/combined_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    ...     loss = model(torch.randn(32, 10)).sum()
+    ...     loss.backward()
+    ...     combined_opt.step()
+    ...     scheduler.step()
+    """
+
+    def __init__(
+        self,
+        optimizers: Sequence[Optimizer],
+        torch_compile_kwargs: dict[str, Any] | None = None,
```
Comment (coreyjadams) [inline physicsnemo/optim/combined_optimizer.py:105]: This syntax is not upstream, right?  I assume we can't put `step` into a compile wrapper since that will break the closure behavior?

File: physicsnemo/optim/combined_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        **kwargs
+            Keyword arguments to pass to each optimizer's ``zero_grad()``.
+            Common kwargs include ``set_to_none`` (bool).
+        """
+        for opt in self.optimizers:
+            opt.zero_grad(*args, **kwargs)
+
+    def step(
+        self, closure: Callable[[], torch.Tensor] | None = None
+    ) -> torch.Tensor | None:
```
Comment (coreyjadams) [inline physicsnemo/optim/combined_optimizer.py]: Upstream returns float | None, not tensor, FYI, in both the closure and `step`:

https://docs.pytorch.org/docs/stable/generated/torch.optim.Optimizer.step.html#torch.optim.Optimizer.step

File: physicsnemo/optim/combined_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+            The loss value returned by the last optimizer, or None if no closure was provided.
+        """
+        loss = None
+        for step_fn in self.step_fns:
+            if closure is None:
+                step_fn()
+            else:
+                res = step_fn(closure)
+                if res is not None:
+                    loss = res
```
Comment (coreyjadams) [inline physicsnemo/optim/combined_optimizer.py:201]: I don't follow this logic - if there are multiple optimizers and a closure, are they expected to return the same value?  Right now we overwrite with the value of the last non-None `res`.

File: test/optim/test_combined_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    opt2 = Adam(model.layer2.parameters(), lr=0.001)
+    return [opt1, opt2]
+
+
+@pytest.fixture
+def combined_optimizer(optimizers):
+    return CombinedOptimizer(optimizers)
+
+
+class TestInitialization:
```
Comment (coreyjadams) [inline test/optim/test_combined_optimizer.py:52]: All of our other tests are more or less "functionals" and not class-based tests.  I don't personally have any objection to this style, but I don't know what unforeseen consequences we might see with this.  Any thoughts?

File: test/optim/test_combined_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        new_combined = CombinedOptimizer([opt1, opt2])
+
+        # Load state
+        new_combined.load_state_dict(state)
+
+        # Verify equality (basic check)
+        # Note: strict equality of state dicts might fail due to weak refs or unrelated keys,
+        # so we check structure matches.
+        new_state = new_combined.state_dict()
+        assert len(new_state["optimizers"]) == len(state["optimizers"])
```
Comment (coreyjadams) [inline test/optim/test_combined_optimizer.py:274]: Can we update this test to check save / restore is getting numerics correct too?  Loading the wrong optimizer state or getting the numbers incorrect would be a pretty frustrating bug in training codes.

File: physicsnemo/optim/combined_optimizer.py (hunk 5 above)
Comment (peterdsharpe) [reply physicsnemo/optim/combined_optimizer.py]: I think the upstream signature is wrong in most cases -- the return type of the model will be a loss represented as a singleton `torch.Tensor`, right? In fact, if you do `loss.item()` to convert `tensor` -> `float`, my hunch is that this would actually break since you can't call `.backward()` on a float (it's detached from the computational graph). I'll test this and do some more research to verify, though.

File: test/optim/test_combined_optimizer.py (hunk 7 above)
Comment (peterdsharpe) [reply test/optim/test_combined_optimizer.py:52]: This just adds grouping for more helpful printouts / test failure reporting - no functional change here.

File: physicsnemo/optim/combined_optimizer.py (hunk 3 above)
Comment (peterdsharpe) [reply physicsnemo/optim/combined_optimizer.py]: Fixed 7ddf8ab4df5b192a8ea694d45529bd9b5549f918

File: test/optim/test_combined_optimizer.py (hunk 8 above)
Comment (peterdsharpe) [reply test/optim/test_combined_optimizer.py:274]: fixed

File: physicsnemo/optim/combined_optimizer.py (hunk 6 above)
Comment (peterdsharpe) [reply physicsnemo/optim/combined_optimizer.py:201]: fixed by adding clarification in docstring; last one takes precedence

File: physicsnemo/optim/combined_optimizer.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/optim/combined_optimizer.py:105]: Deliberate design choice to support compilation through the optimizer here. `torch.compile`-ing the step does not break closure behavior.
